src/surfaceqg.jl

```diff
@@ -0,0 +1,370 @@
+  sol, grid = prob.sol, prob.grid
+  return 1 / (grid.Lx * grid.Ly) * parsevalsum(abs2.(sol), grid)
+end
+
+"""
+    buoyancy_dissipation(prob)
+
+Returns the domain-averaged dissipation rate of surface buoyancy variance due
+to small scale diffusion/viscosity. nν must be >= 1.
+
+In an SQG flow this is identical to twice the rate of kinetic energy dissipation
+"""
+@inline function buoyancy_dissipation(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+  @. vars.uh = 2*grid.Krsq^params.nν * abs2(sol)
+  CUDA.@allowscalar vars.uh[1, 1] = 0
+  return params.ν / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+"""
+    buoyancy_work(prob)
+    buoyancy_work(sol, v, grid)
+
+Returns the domain-averaged rate of work of buoyancy variance by the forcing Fh.
+"""
+@inline function buoyancy_work(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+  @. vars.uh =  sol * conj(vars.Fh)
+  return 1 / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+@inline function buoyancy_work(sol, vars::StochasticForcedVars, grid)
+  @. vars.uh =  (vars.prevsol + sol) / 2 * conj(vars.Fh) # Stratonovich
+  # @. vars.uh = grid.invKrsq * vars.prevsol * conj(vars.Fh)           # Ito
+  return 1 / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+@inline work(prob) = work(prob.sol, prob.vars, prob.grid)
+
+"""
+    buoyancy_drag(prob)
+
+Returns the extraction of domain-averaged buoyancy variance by drag/hypodrag μ.
+"""
+@inline function buoyancy_drag(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+  @. vars.uh = grid.Krsq^params.nμ * abs2(sol)
+  vars.uh[1, 1] = 0
+  return params.μ / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+"""
+    buoyancy_advection(prob)
+
+Returns the average of domain-averaged advection of buoyancy variance. This should
+be zero if buoyancy variance is conserved.
+"""
+@inline function buoyancy_advection(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+
+  @. vars.bh = sol
+  @. vars.uh =   im * grid.l  * sqrt.(grid.invKrsq) * sol
+  @. vars.vh = - im * grid.kr * sqrt.(grid.invKrsq) * sol
+
+  ldiv!(vars.u, grid.rfftplan, vars.uh)
+  ldiv!(vars.v, grid.rfftplan, vars.vh)
+  ldiv!(vars.b, grid.rfftplan, vars.bh)
+
+  @. vars.u *= vars.b # u*b
+  @. vars.v *= vars.b # v*b
+
+  mul!(vars.uh, grid.rfftplan, vars.u) # \hat{u*b}
+  mul!(vars.vh, grid.rfftplan, vars.v) # \hat{v*b}
+
+  @. vars.bh = ( - im * grid.kr * vars.uh - im * grid.l * vars.vh ) * conj( vars.bh )
+
+  return 1 / (grid.Lx * grid.Ly) * parsevalsum(vars.bh, grid)
+end
+
+"""
+    kinetic_energy_advection(prob)
+
+Returns the average of domain-averaged advection of kinetic energy by the
+leading-order (geostrophic) flow.
+"""
+@inline function kinetic_energy_advection(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+
+  @. vars.b = vars.u * vars.u
+  mul!(vars.bh, grid.rfftplan, vars.u)
+  sol = vars.bh
```

Comment:
This line re-binds the name `sol` to `vars.bh`. It does not copy the data in `bh` into the memory space occupied by `prob.sol`.

Reply:
I think that's OK but 2 lines **below** should be ` mul!(vars.bh, grid.rfftplan, vars.v)`?

Reply:
There were a bunch of errors in the advection diagnostics. I think I fixed them and I added some comments to describe what was going on in each step. I tried these diagnostics in the example and it was encouraging; buoyancy variance advection is very small (conserved!), but the kinetic energy advection is large (non-conserved due to ageostrophic effects)

I had to bring sol in because the advection operator is the sum of several terms. I needed a scratch variable that was the right size to be multiplied by some function of velocities (vars.uh and vars.vh) and then added to the advection (stored as vars.bh)  as each term was calculated. If you can think of a better way to do this I'm all ears!

Reply:
Well, just two concerns: the line this comment refers to does not copy the data in `vars.bh` to `sol`. That would require `sol .= vars.bh` (note the `.` before the `=`). Instead, this line creates a new "name" called `sol` (therefore discarding the original reference to `sol`, which is an alias for `vars.bh`. I suspect this is not what was intended.

The second issue is that modifying `sol` causes the model state to be lost, since `sol` keeps track of the model state (the state of the model is not kept anywhere else).

Two alternative solutions may be used: 1) add a new temporary variable to `vars` to be used for this diagnostic. 2) I am not certain, but because the object being calculated has two terms, I think that three temporary variables may be adequate. One strategy you could use is to store the first term in this diagnostic in `vh` (for example), and then use the remaining two temporary variables (possibly overwriting their previous data) to compute the second term, adding that contribution to the already calculated component stored in `vh`. Hope that makes sense.

Reply:
I fully agree; `sol` should not be used as scratch variable.
@BrodiePearson, how about this:

```Julia
 @inline function kinetic_energy_advection(prob)
  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid

  updatevars!(prob)

  mul!(vars.vh, grid.rfftplan, vars.u.^2)
  @. vars.bh = - im * grid.kr * vars.vh * conj(vars.uh)  # - FFT[∂(u²)/∂x]*conj(uh)
  mul!(vars.vh, grid.rfftplan, vars.u .* vars.v)
  @. vars.bh -= im * grid.l * vars.vh * conj(vars.uh)    # - FFT[∂(uv)/∂y]*conj(uh)

  @. vars.vh = - im * grid.kr * sqrt(grid.invKrsq) * sol # because vars.vh was used as scratch

  mul!(vars.uh, grid.rfftplan, vars.u .* vars.v)
  @. vars.bh -= im * grid.kr * vars.uh * conj(vars.vh)    # - FFT[∂(uv)/∂x]*conj(vh)
  mul!(vars.uh, grid.rfftplan, vars.v.^2)
  @. vars.bh -= im * grid.l * vars.uh * conj(vars.vh)     # - FFT[∂(v²)/∂y]*conj(uh)

  return 1 / (grid.Lx * grid.Ly) * parsevalsum(vars.bh , grid)
end
```

I'm not sure if these are the terms you want to compute; I tried to do what the comments on the `kinetic_energy_advection(prob)` function were suggesting.

Notice how I call `updatevars!(prob)` in the beginning to make sure my variables have what they supposed to have. Also, I need to update `vars.vh` after the first half of the computation is done?